Initial draft of prex integration with PWA (#3063)
magento/pwa-studio: pwa-devdocs/src/product-recs/index.md

When you create a recommendation unit in the Admin UI panel, you will need to place components that render product recommendations on appropriate storefront pages. You will do this only once per [supported page type](https://docs.magento.com/user-guide/marketing/product-recommendations.html#supportedrecs).

## Render recommendations

To control the recommendation unit placement on your PWA storefront, use the local intercept file. See the [extensibility framework][] documentation to learn more. To render the recommendation unit on your PWA storefront, use one of the methods described below.

### Venia UI component

The `Recommendations` component is part of the [venia-product-recommendations][] package.
It contains React components that do the following:

-  Collect and send behavioral data to Adobe Sensei
-  Fetch recommendations from the recommendations service
-  Render the recommendation unit to your storefront page

#### Example

```jsx
//Render recommendations using visual component
import { VeniaProductRecommendations, PageTypes } from "@magento/venia-product-recommendations"

> return <Recommendations pageType={CMS} />

Reviewer: It should be

return <VeniaProductRecommendations pageType={PageTypes.CMS} />

```
### Fetch data only

If you are not using the `venia-ui` package, you can call the recommendations service and receive a JSON payload.
#### Example

```js
import { Pagetypes, useRecsData, VeniaProductRecommendations } from "@magento/venia-product-recommendations"
const {data, error, isLoading} = useRecsData({pagetype: Pagetypes.CMS})

> ```

Reviewer: perhaps add this as well:
```js
import { Pagetypes, useRecsData, VeniaProductRecommendations } from "@magento/venia-product-recommendations"

const {data, error, isLoading} = useRecsData({pagetype: Pagetypes.CMS})
console.log(data.results) // [{unitId: "a4h4j4kd9f-a83u3....]
```

Author: STET as per @abrhim .

1. You can install the PWA (_name of this module_) module from NPM:

   ```sh
   npm install @magento/recommendations-pwa
   ```
   This module installs the storefront events that track and collect behavioral data.

1. To install the Magento Product Recommendations module, see the [Magento developer documentation](https://devdocs.magento.com/recommendations/install-configure.html).
This module expands Magento's existing GraphQL API to include fields that are used to fetch the recommendations from the Recommendations Service.

## Create recommendation units

Creating a product recommendation unit for your PWA storefront is similar to [creating one for a Magento storefront](https://docs.magento.com/user-guide/marketing/create-new-rec.html).
The difference is that after you create a recommendation unit, you then need to fetch that recommendation unit from your PWA storefront.

## Fetch recommendations

There are three ways you can fetch product recommendations from your PWA storefront.
If you are using the Venia theme, Magento provides a component that contains the code used to track and collect behavioral events as well as render the recommendation units on your page.

> If you are not using the Venia theme, there are other options available to you.

Reviewer 2: ```suggestion
If you are not using Venia UI components in your storefront project, there are other options available to you.
```

> ### Venia component (recommended)

Reviewer 2: ```suggestion
### Venia UI component (recommended)
```

> The `Recommendations` component is part of the [venia-ui package]: {link technologies/overview/#custom-react-hooks-and-component %} and contains React components that collect and send behavioral data from your storefront to Adobe Sensei, fetch recommendations from the recommendations service, and render the recommendation unit on your PWA storefront page.

Reviewer 2: ```suggestion
The `Recommendations` component is part of the [venia-ui package]: {link technologies/overview/#custom-react-hooks-and-component %}. 
It contains React components that do the following:
-   Collect and send behavioral data to Adobe Sensei
-   Fetch recommendations from the recommendations service
-   Render the recommendation unit to your storefront page
```

#### Example

> ```js

Reviewer 2: ```suggestion
```jsx
```

1. To install the Magento Product Recommendations backend module, see the [Magento developer documentation](https://devdocs.magento.com/recommendations/install-configure.html).
This module expands Magento's existing GraphQL API to include fields that are used to fetch the recommendations from the Recommendations Service.

## Create recommendation units

Creating a product recommendation unit for your PWA storefront is similar to [creating one for a Magento storefront](https://docs.magento.com/user-guide/marketing/create-new-rec.html).
The difference is that after you create a recommendation unit, you then need to add code to your PWA storefront to explicitly fetch the recommendation unit from the Magento backend. You can use one of the methods described below.

> ## Fetch recommendations

Reviewer: ## Rendering Recommendations

> 

Reviewer: Add this, or something to the effect. 

Due to the limited scope of PWA-Studio's extensibility framework for extensions, this extension is unable to inject recommendations into the markup by itself. We recommend using the local intercept file that storefronts have available to inject Product Recommendations onto the page. Soon we will provide an example of how to inject product recommendations via the local intercept file, but for now please reference this existing [documentation](https://magento.github.io/pwa-studio/pwa-buildpack/extensibility-framework/#targetables-in-storefront-projects) on how to do it. Without injecting product recommendations in the markup, recommendations will not be fetched or rendered.

At this point, the merchant can create and manage recommendation units from the Magento Admin UI then fetch those product recommendation units from their PWA storefront.

## Install the Product Recommendations module

Product Recommendations support on PWA requires installing the `venia-product-recommendations` package and the Product Recommendations Magento module.

1. You can install the PWA `venia-product-recommendations` package from the NPM registry:

   ```sh
   npm install @magento/venia-product-recommendations
   ```
   This package contains everything you need to collect behavioral data and render the recommendations. Some recommendation types use behavioral data from your shoppers to train machine learning models that build personalized recommendations. Other recommendation types use catalog data only and do not use any behavioral data. See the [Magento user guide](https://docs.magento.com/user-guide/marketing/product-recommendations.html#trainmlmodels) to learn how Adobe Sensei trains machine learning models that results in higher quality recommendations.

> 1. To install the Magento Product Recommendations backend module, see the [Magento developer documentation](https://devdocs.magento.com/recommendations/install-configure.html).

Reviewer 3: ```suggestion
1. The backend functionality is provided by the [Magento Product Recommendations module](https://devdocs.magento.com/recommendations/install-configure.html).
```

> 1. To install the Magento Product Recommendations backend module, see the [Magento developer documentation](https://devdocs.magento.com/recommendations/install-configure.html).

Author: Applied @misha-kotov edits which rewrote this section.

> 1. Additionally, you need to install the `module-data-services-graphql` module that expands Magento's existing GraphQL API to include fields that are used to fetch the recommendations from the Recommendations Service.

Reviewer 3: ```suggestion
1. Additionally, you need to install the `module-data-services-graphql` module that expands Magento's existing GraphQL coverage to include fields required for storefront behavioral data collection.
```

> 1. Additionally, you need to install the `module-data-services-graphql` module that expands Magento's existing GraphQL API to include fields that are used to fetch the recommendations from the Recommendations Service.

Author: I updated with the composer syntax

   ```sh
   npm install @magento/module-data-services-graphql
   ```

## Create recommendation units

> Creating a product recommendation unit for your PWA storefront is similar to [creating one for a Magento storefront](https://docs.magento.com/user-guide/marketing/create-new-rec.html).

Reviewer 3: ```suggestion
Creating a product recommendation unit for your PWA storefront is the same as [creating one for a Magento storefront](https://docs.magento.com/user-guide/marketing/create-new-rec.html).
```

> The difference is that after you create a recommendation unit, you then need to add code to your PWA storefront to explicitly render the recommendation unit from the Magento backend.

Reviewer 3: ```suggestion
Once you create a recommendation unit in the admin panel, you will need to place components that render product recommendations on appropriate storefront pages. You will do this only once per [supported page type](https://docs.magento.com/user-guide/marketing/product-recommendations.html#supportedrecs). 
```

## Render recommendations

To control the recommendation unit placement on your PWA storefront, use the local intercept file. See the [extensibility framework][] documentation to learn more. To render the recommendation unit on your PWA storefront, use one of the methods described below.

### Venia UI component

> The `Recommendations` component is part of the [venia-ui package][].

Reviewer 2: Is this accurate? I don't see it listed in the [venia-ui library](https://github.com/magento/pwa-studio/tree/develop/packages/venia-ui/lib/components)

Author: @abrhim is this not updated yet?

Reviewer: it is not part of the venia-ui package, rather it is from the venia-product-recommendations package.

import { VeniaProductRecommendations } from "@magento/venia-product-recommendations"

It contains React components that do the following:

-  Collect and send behavioral data to Adobe Sensei
-  Fetch recommendations from the recommendations service
-  Render the recommendation unit to your storefront page

#### Example

```jsx
//Render recommendations using visual component

> return <Recommendations pageType={CMS} />

Reviewer 2: What does the import statement for this component look like?

Author: @abrhim can you take a look?

Reviewer: import { VeniaProductRecommendations, PageTypes } from "@magento/venia-product-recommendations"

return <VeniaProductRecommendations pageType={PageTypes.CMS} />

```
### Fetch data only

If you are not using the `venia-ui` package, you can call the recommendations service and receive a JSON payload.
#### Example

```js
import { Pagetypes, useRecsData, VeniaProductRecommendations } from "@magento/venia-product-recommendations"
const recs = useRecsData({pagetype: Pagetypes.CMS}) 
// or
return <VeniaProductRecommendations pageType={Pagetypes.PRODUCT} />

> ```

Reviewer 2: `// or` indicates these are two examples, they should be split into two different code blocks with explanations on what each does

Reviewer: <VeniaProductRecommendations> actually renders the recommendations and as such doesn't belong to this section. The only method to fetch data only is to use the `useRecsData` hook. So change the code snippet to: 

```js
import { Pagetypes, useRecsData, VeniaProductRecommendations } from "@magento/venia-product-recommendations"
const {data, error, isLoading} = useRecsData({pagetype: Pagetypes.CMS}) 
```

> After installation, the `@magento/venia-product-recommendations` component will collect required shopper behavior on the storefront with no additional setup. In order to place the recommendation unit on the page, place the `VeniaProductRecommendations` component into the local intercept file for the page where you want the recommendation unit to appear.

Reviewer 2: ```suggestion
After you install the `@magento/venia-product-recommendations` package, your storefront will collect shopper behavior with no additional setup.
To place recommendation items on a page, use the Targetables feature in the local intercept file to insert the `VeniaProductRecommendations` component inside the UI component where you want the recommendation items to appear.
```

---
title: Integrating Product Recommendations
---

You can integrate Product Recommendations powered by [Adobe Sensei](https://www.adobe.com/sensei.html) into your PWA Studio storefront.

## Integration overview

![Product Recommendations for PWA Studio](images/pwa-arch-diag-sensei.svg)

Magento's [Product Recommendations powered by Adobe Sensei](https://docs.magento.com/user-guide/marketing/product-recommendations.html) is a feature backed by several SaaS services.
The **Store** side includes your PWA storefront, which contains the event collector and recommendations layout template, and the backend, which includes the GraphQL endpoints, SaaS Export module, and the Admin UI.

> After you install the Product Recommendations PWA extension on your store, it will start sending [behavioral data](https://devdocs.magento.com/recommendations/events.html) to Adobe Sensei.

Reviewer: Suggested change: 

After you install the Product Recommendations PWA extension on your store, it will start sending [behavioral data](https://devdocs.magento.com/recommendations/events.html) to Adobe Sensei with no additional setup.

### Venia UI component

The `Recommendations` component is part of the [venia-ui package][].
It contains React components that do the following:

-  Collect and send behavioral data to Adobe Sensei
-  Fetch recommendations from the recommendations service
-  Render the recommendation unit to your storefront page

#### Example

```jsx
//Render recommendations using visual component
return <Recommendations pageType={CMS} />
```
### Fetch data only

If you are not using the `venia-ui` package, you can call the recommendations service and receive a JSON payload.
#### Example

```js
import { Pagetypes, useRecsData, VeniaProductRecommendations } from "@magento/venia-product-recommendations"
const recs = useRecsData({pagetype: Pagetypes.CMS}) 
// or
return <VeniaProductRecommendations pageType={Pagetypes.PRODUCT} />
```

After you install the `@magento/venia-product-recommendations` package, your storefront will collect shopper behavior with no additional setup.

> To place recommendation items on a page, use the Targetables feature in the local intercept file to insert the `VeniaProductRecommendations` component inside the UI component where you want the recommendation items to appear.

Author: ```suggestion
To place recommendation units on a page, use the Targetables feature in the local intercept file to insert the `VeniaProductRecommendations` component inside the UI component where you want the recommendation items to appear.
```

```js
import { VeniaProductRecommendations } from '@magento/venia-product-recommendations'

> ```

Reviewer: Remove this section as it is redundant.

### Wrapper functions

To abstract the raw JSON, you can use wrapper functions that collect and send behavioral data from your storefront to Adobe Sensei and fetch recommendation units from the Recommendations service.
#### Example

```js
//recommendations JSON
const recsData = useRecsData({pageType: CMS})
```

[venia-ui package]: <{%link technologies/overview/index.md %}#custom-react-hooks-and-component>

> [extensibility framework]: <{%link pwa-buildpack/extensibility-framework/index.md %}#intercept-files>

Reviewer: This section can be deleted as it is redundant with the above section
